### ISSUE TITLE
Add predict_logits to TabICLClassifier

### ISSUE THREAD

**Issue by NegatedObjectIdentity** [soda-inria/tabicl#126]
Thank you very much for our great work! Really appreciating it! 
I was looking for an option to output logits or log odds with the TabICLClassifier. Similar to decision_function in logistic regression in scikit learn. This is important since explaining the feature contributions in TabICLClassifiers with shap/shapiq in logits isolates true feature interactions driven by the model's logic. 
In contrariety, probability output (predict_proba) introduces mathematical interaction artifacts by the probability boundary. E.g. if two features push a prediction toward 100%, their combined probability cannot exceed 1.0. shap/shapiq will interpret this mathematical capping as a negative interaction between the features, even if the model treats them as completely independent internally.
Hence, I propose to implement a TabICLClassifier.predict_logits(X) similar to scikit learn or TabPFN. Thank you in advance!

**Comment by dholzmueller** [comment]
Hi, I don't fully understand what the decision function should output. If it's just log-odds it would be log(p(class 1)/p(class 0)), which can be computed from predict_proba. Do you just want to have it explicitly in a decision_function to make it easier to use in a pipeline?

**Comment by NegatedObjectIdentity** [comment]
Sorry for my clumsy description. This is probably only relevant for the interpretability case. When doing shap/shapiq feature attribution with predict_proba you will find interactions that are purely due too how the probabilities are calculated, hence, the normalization for capping to 1 will show up as interactions. And since the logit transformation is non-linear you cannot transform back shap values from the probability space to the logit space without losing the additivity of shap values. Therefore one should do shap/shapiq calculations in the logit space to avoid the introduction of the interactions. After that you can do the transform to probabilities. Please let me know if this helps!

**Comment by NegatedObjectIdentity** [comment]
In the shap documentation (https://shap.readthedocs.io/en/latest/example_notebooks/overviews/An%20introduction%20to%20explainable%20AI%20with%20Shapley%20values.html)  in the section: Explaining a linear logistic regression model you find all the details.
TLDR;
_If we use SHAP to explain the probability of a linear logistic regression model we see strong interaction effects. This is because a linear logistic regression model is NOT additive in the probability space._
_If we instead explain the log-odds output of the model we see a perfect linear relationship between the models inputs and the model’s outputs. It is important to remember what the units are of the model you are explaining, and that explaining different model outputs can lead to very different views of the model’s behavior._
As far as I understand, this applies in general to all models with that logits to probabilities transform in the end (although I am very aware that TabICL can use interactions and is not a linear model). Hence, to understand how a model works we need to use logits, before normalizing for probabilities.

Furthermore in the shapiq logic at https://github.com/mmschlk/shapiq/blob/main/src/shapiq/explainer/utils.py#L202 you will find a heuristic of which prediction function should be used. predict_logits is the highest ranked one since this is how shap values should be done in most of the cases. The other options (predict_proba, predict) are used as fallbacks.

**Comment by NegatedObjectIdentity** [comment]
<img width="3138" height="2273" alt="Image" src="https://github.com/user-attachments/assets/02ab6c1b-fcc2-40d6-b705-8e85e4201bae" />

<img width="3138" height="2273" alt="Image" src="https://github.com/user-attachments/assets/895a50f6-d202-436b-a19f-fe65fc564ad9" />
Here is a comparison of using tabicl.predict_proba for shapiq (top, probability based) and tabicl.predict_logit for shapiq (bottom, predict_logit) an the drug dataset (figures show shap values for class 0). The values are naturally different (logt vs probabilites). However, the ranking should be the same. However, interactions are artificially changed due to the probability calculation. One can see that in the probabilities case, interaction is in position one. In the logit case, interaction is only third.

**Comment by NegatedObjectIdentity** [comment]
To make the above plots I split the predict_proba from tabicl classifier into two functions. predict_logit and a new predict_proba that makes use of predict_logit like predict makes use of predict_proba. See below:

def predict_logits(self, X: np.ndarray) -> np.ndarray:
        """Predict aggregated class logits for test samples.

        Applies the ensemble of TabICL models to make predictions, with each ensemble
        member providing predictions that are then averaged. The method:

        1. Transforms input data using the fitted encoders
        2. Applies the ensemble generator to create multiple views
        3. Forwards each view through the model
        4. Corrects for class shuffles
        5. Averages predictions across ensemble members

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Test samples for prediction.  Columns that are entirely NaN are
            treated as masked features and excluded from inference.  This is
            useful for computing SHAP values, where masked features are
            represented as all-NaN columns.

        Returns
        -------
        np.ndarray of shape (n_samples, n_classes)
            Class logits for each test sample.
        """
        check_is_fitted(self)
        if isinstance(X, np.ndarray) and len(X.shape) == 1:
            # Reject 1D arrays to maintain sklearn compatibility
            raise ValueError(f"The provided input X is one-dimensional. Reshape your data.")

        # Check if prediction is possible
        has_kv_cache = hasattr(self, "model_kv_cache_") and self.model_kv_cache_ is not None
        has_training_data = (
            has_weakref := hasattr(self, "ensemble_generator_") and getattr(self.ensemble_generator_, "X_", None) is not None
        )
        if not has_kv_cache and not has_training_data:
            raise RuntimeError(
                "Cannot predict: this estimator was saved without training data and has no KV cache. "
                "Predictions require either cached KV projections or the original training data. "
                "Re-fit the estimator or load from a file saved with save_training_data=True or "
                "save_kv_cache=True."
            )

        if self.n_jobs is not None:
            assert self.n_jobs != 0
            old_n_threads = torch.get_num_threads()
            n_logical_cores = mp.cpu_count()

            if self.n_jobs > 0:
                if self.n_jobs > n_logical_cores:
                    warnings.warn(
                        f"TabICL got n_jobs={self.n_jobs} but there are only {n_logical_cores} logical cores available."
                        f" Only {n_logical_cores} threads will be used."
                    )
                n_threads = max(n_logical_cores, self.n_jobs)
            else:
                n_threads = max(1, mp.cpu_count() + 1 + self.n_jobs)

            torch.set_num_threads(n_threads)

        # Preserve DataFrame structure to retain column names and types for correct feature transformation
        X = validate_data(self, X, reset=False, dtype=None, skip_check_array=True)

        # Detect all-NaN columns (used by SHAP's feature masking approach)
        if hasattr(X, "columns"):  # check for dataframe without importing pandas
            feature_mask = X.isna().all(axis=0).to_numpy()
        else:
            arr = np.asarray(X)
            if np.issubdtype(arr.dtype, np.number):
                feature_mask = np.isnan(arr).all(axis=0)
            else:
                # object dtype: v != v is True only for NaN in IEEE 754, safe for strings too
                feature_mask = np.array([all(v != v for v in arr[:, i]) for i in range(arr.shape[1])])

        if feature_mask is not None and not np.any(feature_mask):
            feature_mask = None

        # Fill masked columns so that transformers don't choke on NaN
        if feature_mask is not None:
            if hasattr(X, "columns"):  # Proxy way to check whether X is a dataframe
                X.iloc[:, feature_mask] = 0.0
            else:
                X[:, feature_mask] = 0.0

        X = self.X_encoder_.transform(X)

        # Skip KV cache when features are masked
        has_kv_cache = hasattr(self, "model_kv_cache_") and self.model_kv_cache_ is not None
        use_cache = has_kv_cache and feature_mask is None

        if use_cache:
            # Cache exists: forward only test data and use the pre-computed cache for training data
            test_data = self.ensemble_generator_.transform(X, mode="test")
            outputs = []
            for norm_method, (Xs_test,) in test_data.items():
                kv_cache = self.model_kv_cache_[norm_method]
                outputs.append(self._batch_forward_with_cache(Xs_test, kv_cache))
            outputs = np.concatenate(outputs, axis=0)
        else:
            # No cache or masked features: forward both training and test data
            data = self.ensemble_generator_.transform(X, mode="both", feature_mask=feature_mask)
            outputs = []
            for norm_method, (Xs, ys) in data.items():
                if feature_mask is None:
                    feature_shuffles = self.ensemble_generator_.feature_shuffles_[norm_method]
                else:
                    feature_shuffles = self.ensemble_generator_.masked_feature_shuffles_[norm_method]

                outputs.append(self._batch_forward(Xs, ys, feature_shuffles))
            outputs = np.concatenate(outputs, axis=0)

        # Extract class shuffle patterns from ensemble generator
        class_shuffles = []
        for shuffles in self.ensemble_generator_.class_shuffles_.values():
            class_shuffles.extend(shuffles)

        # Determine actual number of ensemble members
        n_estimators = len(class_shuffles)

        # Aggregate predictions from all ensemble members, correcting for class shuffles
        avg = np.zeros_like(outputs[0])
        for i, shuffle in enumerate(class_shuffles):
            out = outputs[i]
            avg += out[..., shuffle]

        # Calculate ensemble average
        avg /= n_estimators

        if self.n_jobs is not None:
            torch.set_num_threads(old_n_threads)

        return avg

    def predict_proba(self, X: np.ndarray) -> np.ndarray:
        """Predict class probabilities for test samples.

        Uses predict_logits to get class logits and returns the class probabilities
        for each sample.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Test samples for prediction.  Columns that are entirely NaN are
            treated as masked features and excluded from inference.  This is
            useful for computing SHAP values, where masked features are
            represented as all-NaN columns.

        Returns
        -------
        np.ndarray of shape (n_samples, n_classes)
            Class probabilities for each test sample.
        """
        avg = self.predict_logits(X)

        # Convert logits to probabilities
        if self.average_logits:
            avg = self.softmax(avg, axis=-1, temperature=self.softmax_temperature)

        # Normalize probabilities
        return avg / avg.sum(axis=1, keepdims=True)

    def predict(self, X: np.ndarray) -> np.ndarray:
        """Predict class labels for test samples.

        Uses predict_proba to get class probabilities and returns the class with
        the highest probability for each sample.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Test samples for prediction.  Columns that are entirely NaN are
            treated as masked features and excluded from inference.  This is
            useful for computing SHAP values, where masked features are
            represented as all-NaN columns.

        Returns
        -------
        array-like of shape (n_samples,)
            Predicted class labels for each test sample.
        """
        proba = self.predict_proba(X)
        y = np.argmax(proba, axis=1)

        return self.y_encoder_.inverse_transform(y)

**Comment by dholzmueller** [comment]
Thanks! I see two options:
- Implement predict_logits by converting the output of predict_proba() to logits "log(probas) + C". This means we need to choose the constant shift C somehow, either by setting C=0 or mean centering.
- Implement predict_proba() as softmax(predict_logits()). In this case, for average_logits=True (i.e., when the estimator logits are averaged), we can use the original "C" of the logits, maybe that will introduce fewer dependencies. For average_logits=False, we will still need to average probabilities and then convert back to logits and introduce C.

Do you have a preference?

**Comment by NegatedObjectIdentity** [comment]
Thank you for getting back to my inquiry! Sorry for the late response, I was on vacation. The thing is, I am not a software developer, but a scientist who uses yours awesome work, hence, I don't have a strong opinion on how to implement this. However, I would usually prefer the most straight forward solution. This is why I did it like I showed above. Basically nothing new to the working code, I literally just split your code predict_proba code into two functions to give access to the logits that are calculated anyway already in your code, just not returned to the user. Hence, I don't know if there is anything that needs to be changed from my code example. If averaging worked before, it should work with my proposed change as well (at least this is what I belief, please correct me if I am wrong). So, you could forge my changes into a pull request directly? Thank you again, really appreciating your work!
